### ISSUE TITLE
Tpot tries the same Pipeline multiple times in population

### ISSUE THREAD

**Issue by ghgr** [rhiever/tpot#303]
I'm running some tests with tpot (_development_ branch) with some additional _print_ in order to view which pipeline is evaluated at each step. I found out that it often tries the same Pipeline (xgboost). Here is my output:

> Optimization Progress:   1%|          | 301/40040 [2:19:51<283:40:31, 25.70s/pipeline]Pipeline(steps=[('xgbregressor', XGBRegressor(base_score=0.5, colsample_bylevel=1, colsample_bytree=1, gamma=0,
>        learning_rate=0.14, max_delta_step=0, max_depth=7,
>        min_child_weight=20, missing=None, n_estimators=100, nthread=-1,
>        objective='reg:linear', reg_alpha=0, reg_lambda=1,
>        scale_pos_weight=1, seed=0, silent=True, subsample=1.0))])
> Score: 1228.471468
> Optimization Progress:   1%|          | 302/40040 [2:20:16<281:32:19, 25.51s/pipeline]Pipeline(steps=[('xgbregressor', XGBRegressor(base_score=0.5, colsample_bylevel=1, colsample_bytree=1, gamma=0,
>        learning_rate=0.14, max_delta_step=0, max_depth=7,
>        min_child_weight=20, missing=None, n_estimators=100, nthread=-1,
>        objective='reg:linear', reg_alpha=0, reg_lambda=1,
>        scale_pos_weight=1, seed=0, silent=True, subsample=1.0))])
> Score: 1228.471468
> Optimization Progress:   1%|          | 303/40040 [2:20:41<279:33:12, 25.33s/pipeline]Pipeline(steps=[('xgbregressor', XGBRegressor(base_score=0.5, colsample_bylevel=1, colsample_bytree=1, gamma=0,
>        learning_rate=0.14, max_delta_step=0, max_depth=7,
>        min_child_weight=20, missing=None, n_estimators=100, nthread=-1,
>        objective='reg:linear', reg_alpha=0, reg_lambda=1,
>        scale_pos_weight=1, seed=0, silent=True, subsample=1.0))])
> Score: 1228.471468
> Optimization Progress:   1%|          | 304/40040 [2:21:05<277:34:51, 25.15s/pipeline]Pipeline(steps=[('xgbregressor', XGBRegressor(base_score=0.5, colsample_bylevel=1, colsample_bytree=1, gamma=0,
>        learning_rate=0.14, max_delta_step=0, max_depth=7,
>        min_child_weight=20, missing=None, n_estimators=100, nthread=-1,
>        objective='reg:linear', reg_alpha=0, reg_lambda=1,
>        scale_pos_weight=1, seed=0, silent=True, subsample=1.0))])
> Score: 1228.471468
> 

You can see that these four individuals belonging to the same generation are exactly the same. In _tpot/operators/regressors/__init__.py I commented all the lines except base and xgboost (I can't afford to have 7/8 cores idle, and xgboost runs nicely in many cores). 
```
from .base import *
#from .elastic_net import *
#from .extra_trees import *
#from .gradient_boosting import *
#from .knnr import *
#from .lasso_lars_cv import *
#from .linear_svr import *
#from .random_forest import *
from .xg_boost_r import *
```

But that doesn't explain why **tpot** chooses repeated individuals in the population. In addition, these calls are not cached, and they run from scratch each time.
What am I missing?

Thanks

**Comment by rhiever** [comment]
TPOT may maintain multiples of the same pipeline in the candidate solution population, but it only evaluates them once (and uses the cached result for the rest). Please let us know if the algorithm seems to be re-evaluating the pipelines multiple times.

As for *why* TPOT does that: Sometimes it's useful in population-based algorithms to keep multiple copies of the most promising solution within the population. This gives the algorithm more copies of the solution to "tinker" with via mutation and crossover, rather than filling that population space with less-fit candidates.

**Comment by ghgr** [comment]
Thanks Randy for your fast reply! It seemed to be re-evaluating the same pipeline each time (that or the cached version took 3 minutes to load). But I changed a few lines of code (like removing Regressors, adding _print_ statements and setting n_jobs to **3** in _cross_val_score_) and thus I don't think it is representative of the current branch (maybe yes, maybe not). In any case I will keep an aye on this issue and if I detect it again in the _main_/_development_ branch I'll let you know with a PoC.

**Comment by weixuanfu2016** [comment]
It showed there are more than 40k pipelines in TPOT pipeline optimization process in the stdout. What is the setting of population size and generation number? I think if the population size is more than 400 with limited number of operators, it is possible that many duplicated pipelines is generated in the creator (generation 0). I guess the initial evaluation may process all the individuals, including these duplicates, in generation 0.

**Comment by ghgr** [comment]
Thanks for answering! That's an interesting insight and could easily explain the problem. I checked but unfortunately my population was just 40 individuals (and 1000 generations).

**Comment by rhiever** [comment]
This issue should be resolved in the upcoming release thanks to @weixuanfu2016's PR.
